aten/src/ATen/native/Activation.cpp

```diff
@@ -43,6 +43,205 @@ Tensor & rrelu_(Tensor & self, Scalar lower, Scalar upper, bool training, Genera
   return at::rrelu_with_noise_(self, self.type().tensor(), lower, upper, training, generator);
 }
 
+// -----------------------------------
+// prelu forward
+// -----------------------------------
+template <typename scalar_t>
+void prelu_cpu_kernel(
+  Tensor& result,
+  const Tensor& input,
+  const Tensor& weight,
+  int64_t input_dim0_size,
+  int64_t channel_size,
+  int64_t input_stride0,
+  int64_t input_stride1) {
+
+  int64_t i, j, k;
+  auto result_data = result.data<scalar_t>();
+  auto input_data = input.data<scalar_t>();
+  auto weight_data = weight.data<scalar_t>();
+
+  // multiply values at each channel with weight[channel_index]
+  #pragma omp parallel for private(i,j,k) if (input.numel() > 10000)
+  for (i = 0; i < input_dim0_size; i++) {
+    for (j = 0; j < channel_size; j++) {
+      for (k = 0; k < input_stride1; k++) {
+        int64_t pos = i * input_stride0 + j * input_stride1 + k;
+        result_data[pos] = (input_data[pos] > 0) ? input_data[pos] : weight_data[j] * input_data[pos];
```

Comment:
This code can be made more efficient by communicating the the fast that speculation is acceptable to the compiler:

> scalar_t weight = weight_data[j];
> scalar_t input = input_data[pos];
> scalar_t scaled_output = weight * input;
> result_data[pos] = (input > 0) ? input : scaled_output;

You can go even farther down the rabbit hole by hoisting the computation of `weight` out a loop level, and by using __restrict__ and the aligned attributed as appropriate.  Maybe not worth the effort.

Reply:
Just making the speculation explicit is sufficient to convince clang to vectorize the loop where the naïve version did  not get vectorized.

Reply:
Thanks a lot! Indeed I should've cached the input_data! I will compress the nested loops into just one loop later on